internetarchive/openlibrary
# Created a Fulltext Suggestion Macro for 0 Results

Closes #1001 


Feature:
Creates a partial that is rendered within [work_search.html ](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/templates/work_search.html) in the case when 0 results are returned from a search. 

The Fulltext-Search-Suggestion partial retrieves data from the Partials endpoint and renders a card with two suggestions of books with fulltext snippets that match the user's query. If the suggested book or title is clicked, it leads to the book's page. If the fulltext snippet or header/footer is clicked, it leads to the Search Inside results page. 

Edge Cases:
- If there are no results for a given Search Tab AND no Search Inside results -> Only the 'No results' message will appear, no Search Inside suggestions
- All Search categories (books, authors, subjects, lists) return a Search Inside suggestion card if there are 0 category results
- All Search categories return a consistent 'No results' message if there are no category results and no Search Inside matches. 

Measuring Effectiveness: 

1. Does the Suggestion Card result in more people using Search Inside?
    - Check stats for Search Inside page to see if there in a jump in traffic after deployment
2. Are people using the Suggestion Card?
     - Added `SearchInsideCard|SeeAllClick` to `a` tags in [FulltextSearchSuggestion.html](https://github.com/merwhite11/openlibrary/blob/1001/feature/fulltext-search-box/openlibrary/macros/FulltextSearchSuggestion.html)
     - Added `SearchInsideCard|BookClick` to `a` tags in [FulltextSuggestionItem.html](https://github.com/merwhite11/openlibrary/blob/1001/feature/fulltext-search-box/openlibrary/macros/FulltextSearchSuggestionItem.html)
     - Added `SearchInsideCard|QuoteClick` to `a` tags in [FulltextSuggestionSnippet.html](https://github.com/merwhite11/openlibrary/blob/1001/feature/fulltext-search-box/openlibrary/macros/FulltextSuggestionSnippet.html)

### Technical

In order to test the card, I added a [json of sample data to conf](https://github.com/jimchamp/openlibrary/compare/master...jimchamp:openlibrary:sandbox-fulltext-search-results) 
I then referenced the sample data in [lending.py](https://github.com/jimchamp/openlibrary/compare/master...jimchamp:openlibrary:sandbox-fulltext-search-results).

I 'uploaded' books through the command line by running these commands:
`docker compose exec -e PYTHONPATH=. web bash`
`./scripts/copydocs.py /books/OL32090209M`
`./scripts/copydocs.py /books/OL24375865M`
`./scripts/copydocs.py /books/OL24204364M`

I did not commit the json of sample data or the addition to lending.py. 

### Testing

Happy case:
- Type in a search to the search bar under 'Books' that will likely return 0 results eg 'a cricket in the night' -> Search Inside suggestions will appear
- Repeat the search in authors, subjects and lists categories -> Search Inside suggestions will appear

Edge case:
- Type in a complete gibberish search in books, authors, search inside, subjects and lists categories -> Consistent 'No results' message will appear, no Search Inside suggestions

### Screenshot
Happy case: A loading indicator will appear beneath 'No results found', and then will be replaced with this card:
![Screenshot 2024-07-18 at 4 08 23 PM](https://github.com/user-attachments/assets/e882591b-d2f1-48e5-a9f6-5e9a81751986)

Edge case: If there are no Search Inside matches, only the 'No results' message will appear
![Screenshot 2024-07-18 at 4 11 15 PM](https://github.com/user-attachments/assets/85c44f68-f04c-4fa3-afcd-1c6e9c39967e)



### Stakeholders

@jimchamp @mekarpeles @cdrini
